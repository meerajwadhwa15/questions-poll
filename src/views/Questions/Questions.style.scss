@import './components/List/List.style.scss';
@import './components/Card/Card.style.scss';
