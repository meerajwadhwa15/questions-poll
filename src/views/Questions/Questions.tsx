import React, { Component } from "react";
import { connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { Title, Loader, NoRecordFound } from "@/components";
import QuestionsList from "./components/List";
import { FetchQuestionsActions } from "@/store/actions";

import {
  StateProps,
  ReducerType,
  ReducerState,
  DispatchProps,
  ComponentProps
} from "@/types/questions";

export const mapStateToProps = ({
  QuestionsReducers: { questions, isFetching }
}: ReducerType): ReducerState => ({
  questions,
  isFetching
});

export const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  fetchQuestions: () => {
    dispatch(FetchQuestionsActions());
  }
});

export type Props = ComponentProps & ReducerState & DispatchProps;

class QuestionsListView extends Component<Props, StateProps> {
  componentDidMount() {
    const { fetchQuestions, questions } = this.props;
    if (fetchQuestions && (!questions || !questions.length)) {
      fetchQuestions();
    }
  }

  render() {
    const { questions, isFetching } = this.props;
    return (
      <div>
        <Title>Questions</Title>
        {isFetching ? (
          <Loader />
        ) : questions && questions.length ? (
          <QuestionsList list={questions} />
        ) : (
          <NoRecordFound>No Questions Found</NoRecordFound>
        )}
      </div>
    );
  }
}

export { QuestionsListView };
export default connect<
  ReducerState,
  DispatchProps,
  ComponentProps,
  ReducerType
>(
  mapStateToProps,
  mapDispatchToProps
)(QuestionsListView);
