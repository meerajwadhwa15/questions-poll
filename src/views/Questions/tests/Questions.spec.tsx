import React from "react";
import { QuestionsListView, mapStateToProps } from "../Questions";
import { create } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

const questionsList = [
  {
    question: "Favourite programming language?",
    published_at: "2020-03-12T12:42:40.310135+00:00",
    url: "/questions/1",
    choices: [
      {
        choice: "Swift",
        votes: 39,
        url: "/questions/1/choices/1"
      }
    ]
  },
  {
    question: "bjbjbjb",
    published_at: "2020-03-21T07:41:42.503695+00:00",
    url: "/questions/5",
    choices: [
      {
        choice: "g",
        votes: 0,
        url: "/questions/5/choices/14"
      },
      {
        choice: "ik",
        votes: 0,
        url: "/questions/5/choices/15"
      }
    ]
  }
];

const props = {
  questions: questionsList,
  isFetching: false,
  fetchQuestions: jest.fn()
}; 

const propsWithNolist = {
  questions: [],
  isFetching: false,
  fetchQuestions: jest.fn()
}; 

describe("Questions Enzyme testing", () => {
  test("Should call fetchQuestions", () => {
    const spyFetchQuestionsAction = jest.spyOn(propsWithNolist, "fetchQuestions");
    const component = create(
      <BrowserRouter>
        <QuestionsListView {...propsWithNolist}  />
      </BrowserRouter>
    );

    const instance = component.getInstance();

    instance.componentDidMount();
    expect(spyFetchQuestionsAction).toBeCalled();
  });

  test("Should not call fetchQuestions", () => {
    const spyFetchQuestionsAction = jest.spyOn(props, "fetchQuestions");
    const component = create(
      <BrowserRouter>
        <QuestionsListView {...props}  />
      </BrowserRouter>
    );

    const instance = component.getInstance();

    instance.componentDidMount();
    expect(spyFetchQuestionsAction).not.toBeCalled();
  });

  test("Matches the snapshot with QuestionsList", () => {
    const component = create(
      <BrowserRouter>
        <QuestionsListView {...props} />
      </BrowserRouter>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Should test mapStateToProps", () => {
    expect(
      mapStateToProps({
        QuestionsReducers: { questions: questionsList, isFetching: false }
      })
    ).toEqual({
      questions: questionsList,
      isFetching: false
    });
  });
});
