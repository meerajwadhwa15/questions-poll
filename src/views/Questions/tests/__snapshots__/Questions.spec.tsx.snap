// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions Enzyme testing Matches the snapshot with QuestionsList 1`] = `
<div>
  <h1
    className="page-title"
  >
    Questions
  </h1>
  <div
    className="questions-list"
  >
    <div
      className="card"
    >
      <a
        href="/questions/1"
        onClick={[Function]}
        rel="noopener"
        title="Favourite programming language?"
      >
        <h3
          className="card__title ellipsis"
        >
          Favourite programming language?
        </h3>
        <p>
          Published Date: 
          <strong>
            12/Jun/2020
          </strong>
        </p>
        <p>
          Choices: 
          <strong>
            1
          </strong>
        </p>
      </a>
    </div>
    <div
      className="card"
    >
      <a
        href="/questions/5"
        onClick={[Function]}
        rel="noopener"
        title="bjbjbjb"
      >
        <h3
          className="card__title ellipsis"
        >
          bjbjbjb
        </h3>
        <p>
          Published Date: 
          <strong>
            21/Jun/2020
          </strong>
        </p>
        <p>
          Choices: 
          <strong>
            2
          </strong>
        </p>
      </a>
    </div>
  </div>
</div>
`;
